pkg/kubelet/kubelet.go

```diff
@@ -3282,7 +3311,21 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
 
 func (kl *Kubelet) generateAPIPodStatus(pod *api.Pod, podStatus *kubecontainer.PodStatus) api.PodStatus {
 	glog.V(3).Infof("Generating status for %q", format.Pod(pod))
+
+	// any pod that requires eviction must be added to the list to sync.
+	attrs := &lifecycle.PodEvictAttributes{Pod: pod}
+	for _, podEvictor := range kl.podEvictors {
+		if result := podEvictor.Evict(attrs); result.Evict {
```

Comment:
@vishh - the WIP PR that adds eviction around memory is here (see manager.go): https://github.com/kubernetes/kubernetes/pull/21274

TBH - I agree that the existing kubelet workflow is complicated even before this PR.  The motivation for this PR is to really get the core pattern identified, but then let unit testing of each discrete pod evictor become much simpler.

There are multiple channels that can all result in the kubelet calling `syncPod`

- pod workers -> from what I can gather, they respond to changes only from the kubelet updates channel (i.e. api server, local manifest file).  each pod worker when its done its work sets an interval defaulting to 1m0s for when the kubelet should doing a sync again (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L221) even if there are no updates from apiserver or filesource.  The actual decision to respond to that wake-up signal is handled outside the pod worker here as part of the `getPodsToSync` flow: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2039 which is invoked every 1s when we tick the syncCh to do a syncLoopIteration https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2386

- syncCh -> this is a timer based loop through all pods that is run every 1s to see if there are pods that require changes.  Basically every 1s, we call this method for all pods and check if we should synch it because a) the pod_workers workqueue has reached its 1m0s timer for what was scheduled before, or b) the pod has exceeded its active deadline, which obviously does not fall neatly at 1m0s intervals, and so 1s fidelity for that seems more appropriate.

So for the purposes of eviction, we now have two things that could cause a pod to be failed pro-actively outside of the updates channel (apiserver/filesource) which pod_workers appear in my view to traditionally derive their work from:
- active deadline
- node stability eviction

Since the pod_worker workqueue is itself read during the syncCh 1s ticker to when each pods resyncInterval (1m0s) is reached, it felt to me that integrating the pod evictor directly with the syncCh flow was proper and was consistent with the precedent set for active deadline.  Going to the pod_worker directly seemed to confuse where it expects to get its work from.

The current active deadline and eviction flow depends on the internal kubelet housekeeping interval which is 2s (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2469) for it to filter out pods whose phase is (failed | succeeded) and kill the pods here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2167.  Some of the TODOs on where/how we should forget pod workers made me think that ultimately, we will depend on this function in some manner to forget things (most notably, the worker)

If we want to not wait for the housekeeping tick to ultimately force kill the pod, I think we can shrink that time by instead having the pod also potentially get killed during the syncPod function here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1743.  I am fine with that if we want to check for a failed pod status in addition to `canRunPod`.

Thoughts?

/cc @yujuhong